test-data/unit/check-custom-plugin.test

```diff
@@ -585,6 +585,67 @@ reveal_type(instance(2))  # N: Revealed type is 'builtins.float*'
 [[mypy]
 plugins=<ROOT>/test-data/unit/plugins/callable_instance.py
 
+[case testGetMethodHooksOnUnions]
+# flags: --config-file tmp/mypy.ini --no-strict-optional
+from typing import Union
+
+class Foo:
+    def meth(self, x: str) -> str: ...
+class Bar:
+    def meth(self, x: int) -> str: ...
+class Other:
+    meth: int
+
+x: Union[Foo, Bar, Other]
+if isinstance(x.meth, int):
+    reveal_type(x.meth)  # N: Revealed type is 'builtins.int'
+else:
+    reveal_type(x.meth(int()))  # N: Revealed type is 'builtins.int'
```

Comment:
Since the plugin switches types between `int`/`str` and the existing types are also `int`/`str`, it's a bit hard to see what is going on. What about using more than two separate types?

Reply:
OK, I now have different return types for two `meth` methods, and correspondingly different type conversions for the two hooks: `str` -> `int`, and `float -> int`, this way the call would fail if a wrong hook would act, and hopefully this is now easier to read.